xgps.1
.\"Generated by db2man.xsl. Don't modify this, modify the source.
.de Sh \" Subsection
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Ip \" List item
.br
.ie \\n(.$>=3 .ne \\$3
.el .ne 3
.IP "\\$1" \\$2
..
.TH "GPS" 1 "" "" ""
.SH NAME
xgps, xgpsspeed, cgps, gpxlogger \- test clients for gpsd
.SH "SYNOPSIS"
.ad l
.hy 0
.HP 5
\fBxgps\fR [\fIX\-options\fR] [\-h] [\-v] [\-speedunits\ {\fBmph\fR\ |\ \fBkph\fR\ |\ \fBknots\fR}] [\-altunits\ {\fBfeet\fR\ |\ \fBmeters\fR}] [\-l\ [\fBd\fR\ |\ \fBm\fR\ |\ \fBs\fR]] [\-s\ \fIsmoothing\fR] [\fIserver\fR [\fI:port\fR [\fI:device\fR]]]
.ad
.hy
.ad l
.hy 0
.HP 10
\fBxgpsspeed\fR [\-rv] [\fIX\-options\fR] [\-h] [\-v] [\-nc\ \fIX\-color\fR] [\-speedunits\ {\fBmph\fR\ |\ \fBkph\fR\ |\ \fBknots\fR}] [\fIserver\fR [\fI:port\fR [\fI:device\fR]]]
.ad
.hy
.ad l
.hy 0
.HP 5
\fBcgps\fR [\-h] [\-s] [\-v] [\-speedunits\ {\fBmph\fR\ |\ \fBkph\fR\ |\ \fBknots\fR}] [\-altunits\ {\fBfeet\fR\ |\ \fBmeters\fR}] [\-l\ [\fBd\fR\ |\ \fBm\fR\ |\ \fBs\fR]] [\fIserver\fR [\fI:port\fR [\fI:device\fR]]]
.ad
.hy
.ad l
.hy 0
.HP 10
\fBgpxlogger\fR [\fIlogfile\fR]
.ad
.hy
.ad l
.hy 0
.HP 11
\fBcgpxlogger\fR [\-s\ \fIgpsd\-server\fR] [\-p\ \fIgpsd\-port\fR] [\-i\ \fIpoll\-interval\fR] [\-h]
.ad
.hy

.SH "DESCRIPTION"

.SS "xgps"

.PP
xgps is a simple test client for gpsd with an X interface\&. It displays current GPS position/time/velocity information and (for GPSes that support the feature) the locations of accessible satellites\&.

.PP
xgps accepts an \-h option as for gpsd, or a \-v option to dump the package version and exit\&.

.PP
An optional argument may specify a server to get data from; a colon\-separated suffix is taken as a port number\&. If there is a second colon\-separated suffix, that is taken as a device name to be handed to the daemon in an F= command\&.

.PP
The \fB\-speedunits\fR option can be used to set the speed units for display; follow the keyword with knots for nautical miles per hour, kph for kilometres per hour, or mph for miles per hour\&. The default is miles per hour\&. This option can also be set as the X resource 'speedunits'\&.

.PP
The \fB\-altunits\fR option can be used to set the altitude units for display; follow the keyword with 'meters' or 'feet'\&. The default is feet\&. This option can also be set as the X resource 'altunits'\&.

.PP
The \fB\-l\fR option sets the format of latitude and longitude reports\&. The value 'd' produces decimal degrees and is the default\&. The value 'm' produces degrees and decimal minutes\&. The value 's' produces degrees, minutes, and decimal seconds\&.

.PP
The \fB\-s\fR option sets the number of received sentences to process after any fix is received before blanking any field\&. By default it is set to 4, which prevents annoying display flicker on NMEA devices that send several difference sentence types per update cycle, only one of which is a true fix\&. Set this option to zero to disable smoothing and see each update exactly as it is sent\&.

.PP
There is a known bug in xgps; it assumes the default font size is no more than 18 pixels\&. If this is not the case, the satellite data display will show fewer than 12 satellites\&.

.SS "xgpsspeed"

.PP
xgpsspeed is a speedometer that uses position information from the GPS\&. It accepts an \-h option and optional argument as for gps, or a \-v option to dump the package version and exit\&. Additionally, it accepts \-rv (reverse video) and \-nc (needle color) options\&.

.PP
The \-speedunits option can be used to set the speed units for display; follow the keyword with knots for nautical miles per hour, kph for kilometres per hour, or mph for miles per hour\&. The default is miles per hour\&. This option can also be set as the X resource 'speedunits'\&.

.SS "cgps"

.PP
cgps is a client resembling xgps, but without the pictorial satellite display and able to run on a serial terminal or terminal emulator\&.

.PP
The \fB\-s\fR option prevents cgps from printing the raw data\&. This display can also be toggled with the s command\&.

.PP
Rather than use X resources to determine which units to use, cgps looks at variables in its environment\&. Here are the variables and values it checks:

.IP

    GPSD_UNITS one of: 
              imperial   = miles/feet
              nautical   = knots/feet
              metric     = km/meters
    LC_MEASUREMENT
	      en_US      = miles/feet
              C          = miles/feet
              POSIX      = miles/feet
              [other]    = km/meters
    LANG
	      en_US      = miles/feet
              C          = miles/feet
              POSIX      = miles/feet
              [other]    = km/meters

.PP
cgps terminates when you send it a SIGHUP or SIGINT; given default terminal settings this will happen when you type Ctl\-C at it\&. It will also terminate on 'q'

.SS "cgpxlogger"

.PP
This program connects to gpsd, polls for location, and logs each fix to standard output in GPX (XML) format\&.

.PP
The program also writes start and end messages to syslog\&.

.SS "gpxlogger"

.PP
This program listens to DBUS broadcasts from gpsd (org\&.gpsd\&.fix) and logs each fix to standard output as they arrive in an XML format\&.

.PP
The output may be composed of multiple tracks\&. A new track is created if there's no fix for 5 seconds\&.

.PP
The program also writes start and end messages to syslog\&.

.SH "SEE ALSO"

.PP
 \fBgpsd\fR(8), \fBlibgps\fR(3), \fBlibgpsd\fR(3), \fBgpsprof\fR(1), \fBgpsfake\fR(1)\&.

.SH "AUTHORS"

.PP
Remco Treffcorn, Derrick Brashear, Russ Nelson & Eric S\&. Raymond(xgps)\&. Jeff Francis (cgps)\&. Amaury Jacquot <sxpert@esitcom\&.org> & Petter Reinholdtsen <pere@hungry\&.com> (gpxlogger)\&. Chris Kuethe <chris\&.kuethe@gmail\&.com> (cgpxlogger)\&.

.PP
This manual page by Eric S\&. Raymond <esr@thyrsus\&.com>\&. There is a project page, with xgps screenshots, at berlios\&.de: \fIhttp://gpsd.berlios.de/\fR\&.



xgps.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <unistd.h>
#include <math.h>
#include <errno.h>
#include <Xm/Xm.h>
#include <Xm/MwmUtil.h>
#include <Xm/PushB.h>
#include <Xm/Form.h>
#include <Xm/RowColumn.h>
#include <Xm/Label.h>
#include <Xm/TextF.h>
#include <Xm/List.h>
#include <Xm/DrawingA.h>
#include <Xm/Protocols.h>
#include <X11/Shell.h>

#include "config.h"
#include "gps.h"
#include "display.h"

/* 
 * Widget and window sizes.
 */
#define MAX_FONTSIZE	18		/* maximum fontsize we handle*/
/* height of satellite-data display */
#define SATDATA_HEIGHT	MAX_FONTSIZE*(MAXCHANNELS+1)
#define LEFTSIDE_WIDTH	205		/* width of data-display side */
#define SATDIAG_SIZE	400		/* size of satellite diagram */

static Widget toplevel, form, left, right;
#ifdef EXPLICIT_QUIT
static Widget quitbutton;
#endif /* EXPLICIT_QUIT */
static Widget satellite_list, satellite_diagram, status;
static Widget rowColumn_11, rowColumn_12, rowColumn_13, rowColumn_14;
static Widget rowColumn_15, rowColumn_16, rowColumn_17, rowColumn_18;
static Widget rowColumn_19, rowColumn_20;
static Widget text_1, text_2, text_3, text_4, text_5;
static Widget text_6, text_7, text_8, text_9, text_10;
static Widget label_1, label_2, label_3, label_4, label_5;
static Widget label_6, label_7, label_8, label_9, label_10;
static GC gc;

/*@ -nullassign @*/
static XrmOptionDescRec options[] = {
{"-altunits",  "*altunits",            XrmoptionSepArg,        NULL},
{"-speedunits","*speedunits",          XrmoptionSepArg,        NULL},
};
String fallback_resources[] = {NULL};
/*@ +nullassign @*/

struct unit_t {
    char *legend;
    double factor;
};
static struct unit_t speedtable[] = {
    {"knots",		1/KNOTS_TO_KPH},
    {"mph",		MPS_TO_MPH},
    {"kph",		MPS_TO_KPH},
}, *speedunits = speedtable;
static struct unit_t  alttable[] = {
    {"feet",		METERS_TO_FEET},
    {"meters",		1},
}, *altunits = alttable;

static void quit_cb(void)
{
    exit(0);	/* closes the GPS along with other fds */
}

/*@ -mustfreefresh -compdef +ignoresigns @*/
static Pixel get_pixel(Widget w, char *resource_value)
{
    Colormap colormap;
    Boolean cstatus;
    XColor exact, color;

    colormap = DefaultColormapOfScreen(DefaultScreenOfDisplay(XtDisplay(w)));
    /*@i@*/cstatus = XAllocNamedColor(XtDisplay(w), colormap, resource_value, &color, &exact);
    if (cstatus == (Boolean)False) {
	(void)fprintf(stderr, "Unknown color: %s", resource_value);
	color.pixel = BlackPixelOfScreen(DefaultScreenOfDisplay(XtDisplay(w)));
    };
    /*@i1@*/return (color.pixel);
}

static void build_gui(Widget toplevel)
{
    Arg args[100];
    XGCValues gcv;
    Atom delw;
    int i;
    XmString string;

    /*@ -immediatetrans -usedef @*/
    /* the root application window */
    XtSetArg(args[0], XmNwidth, LEFTSIDE_WIDTH + SATDIAG_SIZE + 26);
    XtSetArg(args[1], XmNheight, SATDATA_HEIGHT + 14*MAX_FONTSIZE + 12);
#ifdef __UNUSED__
    XtSetArg(args[2], XmNresizePolicy, XmRESIZE_NONE);
    XtSetArg(args[3], XmNallowShellResize, False);
    XtSetArg(args[4], XmNdeleteResponse, XmDO_NOTHING);
    XtSetArg(args[5], XmNmwmFunctions,
	     MWM_FUNC_RESIZE | MWM_FUNC_MOVE | MWM_FUNC_MINIMIZE | MWM_FUNC_MAXIMIZE);
    /*@ +immediatetrans +usedef @*/
    XtSetValues(toplevel, args, 6);
#else
    /*@ +immediatetrans +usedef @*/
    XtSetValues(toplevel, args, 2);
#endif /* __UNUSED__ */

    /*@ -onlytrans @*/
    /* a form to assist with geometry negotiation */
    form = XtVaCreateManagedWidget("form", xmFormWidgetClass, toplevel, NULL);
    /* the left half of the screen */
    left = XtVaCreateManagedWidget("left", xmRowColumnWidgetClass, form,
				   XmNleftAttachment, XmATTACH_FORM,
				   XmNtopAttachment, XmATTACH_FORM,
				   NULL);
    /* the right half of the screen */
    right = XtVaCreateManagedWidget("right", xmRowColumnWidgetClass, form,
				    XmNleftAttachment, XmATTACH_WIDGET,
				    XmNleftWidget, left,
				    XmNtopAttachment, XmATTACH_FORM,
				    NULL);
    /* the application status bar */
    status = XtVaCreateManagedWidget("status", xmTextFieldWidgetClass, form,
				     XmNcursorPositionVisible, False,
				     XmNeditable, False,
				     XmNmarginHeight, 1,
				     XmNhighlightThickness, 0,
				     XmNshadowThickness, 2,
				     XmNleftAttachment, XmATTACH_FORM,
				     XmNrightAttachment, XmATTACH_FORM,
				     XmNtopAttachment, XmATTACH_WIDGET,
				     XmNtopWidget, left,
				     NULL);
    /* satellite location and SNR data panel */
    satellite_list =
      XtVaCreateManagedWidget("satellite_list", xmListWidgetClass, left,
			      XmNbackground, get_pixel(toplevel, "snow"),
			      XmNheight, SATDATA_HEIGHT,
			      XmNwidth, LEFTSIDE_WIDTH,
			      XmNlistSizePolicy, XmCONSTANT,
			      XmNhighlightThickness, 0,
			      XmNlistSpacing, 4,
			      NULL);
    /* the satellite diagram */
    satellite_diagram = 
      XtVaCreateManagedWidget("satellite_diagram",
			      xmDrawingAreaWidgetClass, right, 
			      XmNbackground, get_pixel(toplevel, "snow"),
			      XmNheight, SATDIAG_SIZE+24, XmNwidth, SATDIAG_SIZE,
			      NULL);
    gcv.foreground = BlackPixelOfScreen(XtScreen(satellite_diagram));
    gc = XCreateGC(XtDisplay(satellite_diagram),
	RootWindowOfScreen(XtScreen(satellite_diagram)), GCForeground, &gcv);
    register_canvas(satellite_diagram, gc);
    /*@i@*/XtAddCallback(satellite_diagram, XmNexposeCallback, (XtPointer)redraw, NULL);
    /* the data display */
    /*@ -immediatetrans @*/
    XtSetArg(args[0], XmNorientation, XmHORIZONTAL);
    /*@ +immediatetrans @*/
    rowColumn_11 = XtCreateManagedWidget("time     ", xmRowColumnWidgetClass, left, args, 1);

    rowColumn_12 = XtCreateManagedWidget("latitude ", xmRowColumnWidgetClass, left, args, 1);
    rowColumn_13 = XtCreateManagedWidget("longitude", xmRowColumnWidgetClass, left, args, 1);
    rowColumn_14 = XtCreateManagedWidget("altitude ", xmRowColumnWidgetClass, left, args, 1);
    rowColumn_15 = XtCreateManagedWidget("speed    ", xmRowColumnWidgetClass, left, args, 1);
    rowColumn_16 = XtCreateManagedWidget("track    ", xmRowColumnWidgetClass, left, args, 1);
    rowColumn_17 = XtCreateManagedWidget("eph      ", xmRowColumnWidgetClass, left, args, 1);
    rowColumn_18 = XtCreateManagedWidget("epv      ", xmRowColumnWidgetClass, left, args, 1);
    rowColumn_19 = XtCreateManagedWidget("climb    ", xmRowColumnWidgetClass, left, args, 1);
    rowColumn_20 = XtCreateManagedWidget("status   ", xmRowColumnWidgetClass, right, args, 1);

    label_1 = XtCreateManagedWidget("Time     ", xmLabelWidgetClass, rowColumn_11, args, 0);
    label_2 = XtCreateManagedWidget("Latitude ", xmLabelWidgetClass, rowColumn_12, args, 0);
    label_3 = XtCreateManagedWidget("Longitude", xmLabelWidgetClass, rowColumn_13, args, 0);
    label_4 = XtCreateManagedWidget("Altitude ", xmLabelWidgetClass, rowColumn_14, args, 0);
    label_5 = XtCreateManagedWidget("Speed    ", xmLabelWidgetClass, rowColumn_15, args, 0);
    label_6 = XtCreateManagedWidget("Course   ", xmLabelWidgetClass, rowColumn_16, args, 0);
    label_7 = XtCreateManagedWidget("EPH      ", xmLabelWidgetClass, rowColumn_17, args, 0);
    label_8 = XtCreateManagedWidget("EPV      ", xmLabelWidgetClass, rowColumn_18, args, 0);
    label_9 = XtCreateManagedWidget("Climb    ", xmLabelWidgetClass, rowColumn_19, args, 0);
    label_10= XtCreateManagedWidget("Status   ", xmLabelWidgetClass, rowColumn_20, args, 0);

    /*@ -immediatetrans @*/
    XtSetArg(args[0], XmNcursorPositionVisible, False);
    XtSetArg(args[1], XmNeditable, False);
    XtSetArg(args[2], XmNmarginHeight, 2);
    XtSetArg(args[3], XmNhighlightThickness, 0);
    XtSetArg(args[4], XmNshadowThickness, 1);
    XtSetArg(args[5], XmNcolumns, 23);
    /*@ +immediatetrans @*/
    text_1 = XtCreateManagedWidget("text_1", xmTextFieldWidgetClass,
				   rowColumn_11, args, 6);
    text_2 = XtCreateManagedWidget("text_2", xmTextFieldWidgetClass,
				   rowColumn_12, args, 6);
    text_3 = XtCreateManagedWidget("text_3", xmTextFieldWidgetClass,
				   rowColumn_13, args, 6);
    text_4 = XtCreateManagedWidget("text_4", xmTextFieldWidgetClass,
				   rowColumn_14, args, 6);
    text_5 = XtCreateManagedWidget("text_5", xmTextFieldWidgetClass,
				   rowColumn_15, args, 6);
    text_6 = XtCreateManagedWidget("text_6", xmTextFieldWidgetClass,
				   rowColumn_16, args, 6);
    text_7 = XtCreateManagedWidget("text_7", xmTextFieldWidgetClass,
				   rowColumn_17, args, 6);
    text_8 = XtCreateManagedWidget("text_8", xmTextFieldWidgetClass,
				   rowColumn_18, args, 6);
    text_9 = XtCreateManagedWidget("text_9", xmTextFieldWidgetClass,
				   rowColumn_19, args, 6);
    text_10 = XtCreateManagedWidget("text_10", xmTextFieldWidgetClass,
				   rowColumn_20, args, 6);

#ifdef EXPLICIT_QUIT
    quitbutton = XtCreateManagedWidget("            Quit            ",
			 xmPushButtonWidgetClass, rowColumn_20, args, 0);
    /*@i@*/XtAddCallback(quitbutton, XmNactivateCallback, (XtPointer)quit_cb, NULL);
#endif /* EXPLICIT_QUIT */

    XtRealizeWidget(toplevel);
    delw = XmInternAtom(XtDisplay(toplevel), "WM_DELETE_WINDOW", False);
    /*@ -nullpass @*//* splint 3.1.1 lacks annotated prototype... */
    /*@i@*/XmAddWMProtocolCallback(toplevel, delw,
			    (XtCallbackProc)quit_cb, (XtPointer)NULL);
    /*@ +onlytrans +nullpass @*/

    /* create empty list items to be replaced on update */
    string = XmStringCreateSimple(" ");
    for (i = 0; i <= MAXCHANNELS; i++)
	XmListAddItem(satellite_list, string, 0);
    XmStringFree(string);
}
/*@ +mustfreefresh -ignoresigns +immediatetrans @*/

static void handle_time_out(XtPointer client_data UNUSED,
			    XtIntervalId *ignored UNUSED)
/* runs when there is no data for a while */
{
    XmTextFieldSetString(status, "no data arriving");
    XmTextFieldSetString(text_10, "UNKNOWN");
}

/*
 * No dependencies on the session structure above this point.
 */

static struct gps_data_t *gpsdata;
static time_t timer;	/* time of last state change */
static int state = 0;	/* or MODE_NO_FIX=1, MODE_2D=2, MODE_3D=3 */
static int smoothing = 4;	/* # of transmitted sentences to smooth across */
static XtAppContext app;
static XtIntervalId timeout;
static enum deg_str_type deg_type = deg_dd;

static void handle_input(XtPointer client_data UNUSED, int *source UNUSED,
			 XtInputId *id UNUSED)
{
    if (gps_poll(gpsdata) < 0) {
	(void)fprintf(stderr, "Read error on server socket.");
	exit(1);
    }
}

static void update_panel(struct gps_data_t *gpsdata, 
			 char *message, 
			 size_t len UNUSED, int level UNUSED)
/* runs on each sentence */
{
    static int lfok = 0;
    unsigned int i;
    int newstate, newtxt;
    XmString string[MAXCHANNELS+1];
    char s[128], *latlon, *sp;

    if (message[0] != '\0')
	while (isspace(*(sp = message + strlen(message) - 1)))
	    *sp = '\0';
    XmTextFieldSetString(status, message);
    /* This is for the satellite status display */
    if (gpsdata->satellites) {
	string[0] = XmStringCreateSimple("PRN:   Elev:  Azim:  SNR:  Used:");
	for (i = 0; i < MAXCHANNELS; i++) {
	    if (i < (unsigned int)gpsdata->satellites) {
		(void)snprintf(s, sizeof(s),  
			       " %3d    %02d    %03d    %02d      %c", 
			       gpsdata->PRN[i],
			       gpsdata->elevation[i], gpsdata->azimuth[i], 
			       gpsdata->ss[i],	gpsdata->used[i] ? 'Y' : 'N');
	    } else
		(void)strcpy(s, "                  ");
	    string[i+1] = XmStringCreateSimple(s);
	}
	XmListReplaceItemsPos(satellite_list, string, (int)sizeof(string), 1);
	for (i = 0; i < (unsigned int)(sizeof(string)/sizeof(string[0])); i++)
	    XmStringFree(string[i]);
    }
    /* here are the value fields */
    if (isnan(gpsdata->fix.time)==0) {
	    (void)unix_to_iso8601(gpsdata->fix.time, s, (int)sizeof(s));
	    newtxt = 1;
	} else {
	    newtxt = (lfok>0) ? 0 : ((void)strcpy(s, "n/a"), 1);
	}
	if (newtxt != 0) XmTextFieldSetString(text_1, s);
	if (gpsdata->fix.mode >= MODE_2D) {
	    latlon = deg_to_str(deg_type,  fabs(gpsdata->fix.latitude));
	    newtxt = snprintf(s, sizeof(s), "%s %c", latlon, (gpsdata->fix.latitude < 0) ? 'S' : 'N');
	} else {
	    newtxt = (lfok>0) ? 0 : ((void)strcpy(s, "n/a"), 1);
	}
	if (newtxt != 0) XmTextFieldSetString(text_2, s);
	if (gpsdata->fix.mode >= MODE_2D) {
	    latlon = deg_to_str(deg_type,  fabs(gpsdata->fix.longitude));
	    newtxt = snprintf(s, sizeof(s), "%s %c", latlon, (gpsdata->fix.longitude < 0) ? 'W' : 'E');
	} else {
	    newtxt = (lfok>0) ? 0 : ((void)strcpy(s, "n/a"), 1);
	}
	if (newtxt != 0) XmTextFieldSetString(text_3, s);
	if (gpsdata->fix.mode == MODE_3D) {
	    newtxt = snprintf(s, sizeof(s), "%f %s",gpsdata->fix.altitude*altunits->factor, altunits->legend);
	} else {
	    newtxt = (lfok>0) ? 0 : ((void)strcpy(s, "n/a"), 1);
	}
	if (newtxt != 0) XmTextFieldSetString(text_4, s);
	if (gpsdata->fix.mode >= MODE_2D && isnan(gpsdata->fix.track)==0) {
	    newtxt = snprintf(s, sizeof(s), "%f %s", gpsdata->fix.speed*speedunits->factor, speedunits->legend);
	} else {
	    newtxt = (lfok>0) ? 0 : ((void)strcpy(s, "n/a"), 1);
	}
	if (newtxt != 0) XmTextFieldSetString(text_5, s);
	if (gpsdata->fix.mode >= MODE_2D && isnan(gpsdata->fix.track)==0) {
	    newtxt = snprintf(s, sizeof(s), "%f degrees", gpsdata->fix.track);
	} else {
	    newtxt = (lfok>0) ? 0 : ((void)strcpy(s, "n/a"), 1);
	}
	if (newtxt != 0) XmTextFieldSetString(text_6, s);
	if (isnan(gpsdata->fix.eph)==0) {
	    newtxt = snprintf(s, sizeof(s), "%f %s", gpsdata->fix.eph * altunits->factor, altunits->legend);
	} else {
	    newtxt = (lfok>0) ? 0 : ((void)strcpy(s, "n/a"), 1);
	}
	if (newtxt != 0) XmTextFieldSetString(text_7, s);
	if (isnan(gpsdata->fix.epv)==0) {
	    newtxt = snprintf(s, sizeof(s), "%f %s", gpsdata->fix.epv * altunits->factor, altunits->legend);
	} else {
	    newtxt = (lfok>0) ? 0 : ((void)strcpy(s, "n/a"), 1);
	}
	if (newtxt != 0) XmTextFieldSetString(text_8, s);
	if (gpsdata->fix.mode == MODE_3D && isnan(gpsdata->fix.climb)==0) {
	    newtxt = snprintf(s, sizeof(s), "%f %s/sec", gpsdata->fix.climb * altunits->factor, altunits->legend);
	} else {
	    newtxt = (lfok>0) ? 0 : ((void)strcpy(s, "n/a"), 1);
	}
	if (newtxt != 0) XmTextFieldSetString(text_9, s);

	if (gpsdata->online == 0) {
	    newstate = 0;
	    (void)snprintf(s, sizeof(s), "OFFLINE");
	    lfok = 0;
	} else {
	    newstate = gpsdata->fix.mode;
	    switch (gpsdata->fix.mode) {
	    case MODE_2D:
		(void)snprintf(s, sizeof(s), "2D %sFIX",(gpsdata->status==STATUS_DGPS_FIX)?"DIFF ":"");
		if (lfok>0) lfok--;
		break;
	    case MODE_3D:
		(void)snprintf(s, sizeof(s), "3D %sFIX",(gpsdata->status==STATUS_DGPS_FIX)?"DIFF ":"");
		lfok = smoothing;
		break;
	    default:
		(void)snprintf(s, sizeof(s), "NO FIX");
		lfok = 0;
		break;
	    }
	}
	if (newstate != state) {
	    timer = time(NULL);
	    state = newstate;
	}
	(void)snprintf(s+strlen(s), sizeof(s)-strlen(s),
		       " (%d secs)", (int) (time(NULL) - timer));
	XmTextFieldSetString(text_10, s);
	draw_graphics(gpsdata);

	XtRemoveTimeOut(timeout);
	timeout = XtAppAddTimeOut(app, 2000, handle_time_out, NULL);
    }

    static char *get_resource(Widget w, char *name, char *default_value)
    {
      XtResource xtr;
      char *value = NULL;

      /*@ -observertrans -statictrans -immediatetrans -compdestroy @*/
      xtr.resource_name = name;
      xtr.resource_class = "AnyClass";
      xtr.resource_type = XmRString;
      xtr.resource_size = (Cardinal)sizeof(String);
      xtr.resource_offset = 0;
      xtr.default_type = XmRImmediate;
      xtr.default_addr = default_value;
      XtGetApplicationResources(w, &value, &xtr, 1, NULL, 0);
      if (value) return value;
      /*@ +observertrans +statictrans +immediatetrans +compdestroy @*/
      /*@i@*/return default_value;
    }

    /*@ -mustfreefresh @*/
    int main(int argc, char *argv[])
    {
	int option;
	char *arg = NULL, *colon1, *colon2, *device = NULL, *server = NULL, *port = DEFAULT_GPSD_PORT;
	char *su, *au;
	char *err_str = NULL;

	/*@ -onlytrans */
	toplevel = XtVaAppInitialize(&app, "xgps", 
				   options, XtNumber(options), 
				   &argc,argv, fallback_resources,NULL);

	/*@ +onlytrans */
	su = get_resource(toplevel, "speedunits", "mph");
	for (speedunits = speedtable; 
	     speedunits < speedtable + sizeof(speedtable)/sizeof(speedtable[0]);
	     speedunits++)
	    if (strcmp(speedunits->legend, su) == 0)
		goto speedunits_ok;
	speedunits = speedtable;
	(void)fprintf(stderr, "xgps: unknown speed unit, defaulting to %s\n", speedunits->legend);
    speedunits_ok:;

	au = get_resource(toplevel, "altunits",   "feet");
	for (altunits = alttable; 
	     altunits < alttable + sizeof(alttable)/sizeof(alttable[0]);
	     altunits++)
	    if (strcmp(altunits->legend, au) == 0)
		goto altunits_ok;
	altunits = alttable;
	(void)fprintf(stderr, "xgps: unknown altitude unit, defaulting to %s\n", altunits->legend);
    altunits_ok:;

	while ((option = getopt(argc, argv, "hl:s:v")) != -1) {
	    switch (option) {
	    case 'v':
		(void)printf("xgps %s\n", VERSION);
		exit(0);
	    case 'l':
		switch ( optarg[0] ) {
		case 'd':
		    deg_type = deg_dd;
		    continue;
		case 'm':
		    deg_type = deg_ddmm;
		    continue;
		case 's':
		    deg_type = deg_ddmmss;
		    continue;
		default:
		    (void)fprintf(stderr, "Unknown -l argument: %s\n", optarg);
		    /*@ -casebreak @*/
		}
	case 's':
	    smoothing = atoi(optarg);
	    break;
	case 'h': default:
	    (void)fputs("usage:  xgps [-hv] [-speedunits {mph,kph,knots}] [-altunits {ft,meters}] [-l {d|m|s}] [-s smoothing] [server[:port:[device]]]\n", stderr);
	    exit(1);
	}
    }
    /*@ -branchstate @*/
    if (optind < argc) {
	arg = strdup(argv[optind]);
	/*@i@*/colon1 = strchr(arg, ':');
	server = arg;
	if (colon1 != NULL) {
	    if (colon1 == arg)
		server = NULL;
	    else
		*colon1 = '\0';
	    port = colon1 + 1;
	    colon2 = strchr(port, ':');
	    if (colon2 != NULL) {
		if (colon2 == port)
		    port = NULL;
	        else
		    *colon2 = '\0';
		device = colon2 + 1;
	    }
	}
	colon1 = colon2 = NULL;
    }
    /*@ +branchstate @*/

    /*@i@*/gpsdata = gps_open(server, port);
    if (!gpsdata) {
	switch ( errno ) {
	case NL_NOSERVICE: 	err_str = "can't get service entry"; break;
	case NL_NOHOST: 	err_str = "can't get host entry"; break;
	case NL_NOPROTO: 	err_str = "can't get protocol entry"; break;
	case NL_NOSOCK: 	err_str = "can't create socket"; break;
	case NL_NOSOCKOPT: 	err_str = "error SETSOCKOPT SO_REUSEADDR"; break;
	case NL_NOCONNECT: 	err_str = "can't connect to host"; break;
	default:             	err_str = "Unknown"; break;
	}
	(void)fprintf( stderr, 
		       "xgps: no gpsd running or network error: %d, %s\n", 
		       errno, err_str);
	exit(2);
    }

    build_gui(toplevel);

    timeout = XtAppAddTimeOut(app, 2000, handle_time_out, app);
    timer = time(NULL);

    gps_set_raw_hook(gpsdata, update_panel);

    if (device) {
	char *channelcmd = (char *)malloc(strlen(device)+3);

	if (channelcmd) {
	    /*@i1@*/(void)strcpy(channelcmd, "F=");
	    (void)strcpy(channelcmd+2, device);
	    (void)gps_query(gpsdata, channelcmd);
	    (void)free(channelcmd);
	}
    }
	
    (void)gps_query(gpsdata, "w+x\n");

    (void)XtAppAddInput(app, gpsdata->gps_fd, 
		  (XtPointer)XtInputReadMask, handle_input, NULL);
    (void)XtAppMainLoop(app);

    (void)gps_close(gpsdata);
    if (arg != NULL)
	(void)free(arg);
    return 0;
}
/*@ +mustfreefresh @*/
